### PR TITLE
[feat] Generalized MM transformer

### PR DESCRIPTION
**Pull request by vedanuj**
Introduces a generalized version of multimodal transformer.

- Extends `BaseTransformer` and builds a modular multimodal transformer model
- Base transformer model can be any model from [transformers](https://github.com/huggingface/transformers)
- Supports n number of modalities
- Adds ability to add separate input, position and segment embeddings for each modality
- text modalities share weight with base transformer model embeddings
- position embeddings use separate embeddings for each modality
- token type/segment embeddings are shared/mean of base transformer token/segment embeddings (if available)

Test:
- Tested with HM, VQA2, OKVQA

### CODE REVIEW

File: mmf/models/mmf_transformer.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+class ImageEncoder(MultiModalEncoderBase):
+    """Extends the MultiModalEncoderBase class which builds the encoder based on
+    the config parameters. We can set the type of image encoder(resnet50, resnet152,
+    resnext50 etc) and other parameters like num of features, type of pooling etc.
+    """
+
+    def __init__(self, config: DictConfig):
+        super().__init__(config)
+
+    def build(self):
```
Comment (apsdehal) [inline mmf/models/mmf_transformer.py:28]: Do we need this extension? This isn't doing anything special on top of existing

File: mmf/models/mmf_transformer.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        self.encoder = self._build_modal_encoder(self.config.image_encoder)
+
+    def forward(self, x: Tensor) -> Tensor:
```
Comment (apsdehal) [inline mmf/models/mmf_transformer.py:31]: I think this needs to be `Type[Tensor]` where Type is import from `typings`.

File: mmf/models/mmf_transformer.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    have their input id, position id and segment id. We generate embeddings of
+    dimension config.hidden_size for each and then first add the three embeddings
+    for each modality to have a modality specific embedding. We then concat the
+    modality specific embeddings to have a joint embedding.
+    """
+
+    def __init__(
+        self,
+        model_config: Dict[str, Any],
+        transformer_config: Dict[str, Any],
```
Comment (apsdehal) [inline mmf/models/mmf_transformer.py:46]: Is this hugging face config?

File: mmf/models/mmf_transformer.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+        transformer: Type[nn.Module],
+    ):
+        super().__init__()
+        self.model_config = model_config
+        self.transformer_config = transformer_config
+        self._build_layers()
+        self._init_weights(transformer)
+
+    def _build_layers(self):
```
Comment (apsdehal) [inline mmf/models/mmf_transformer.py:55]: We should expose this as this won't fit all use cases.

File: mmf/configs/models/mmf_transformer/defaults.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+model_config:
```
Comment (apsdehal) [inline mmf/configs/models/mmf_transformer/defaults.yaml:1]: We should point to ConfigSchema definition here.

File: mmf/models/mmf_transformer.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+                self,
+                modality + "_dropout",
+                nn.Dropout(self.transformer_config.hidden_dropout_prob),
+            )
+
+        self.token_type_embeddings = nn.Embedding(
+            len(self.model_config.modalities), self.transformer_config.hidden_size
+        )
+
+    def _init_weights(self, transformer: Type[nn.Module]):
```
Comment (apsdehal) [inline mmf/models/mmf_transformer.py:106]: It would make sense to expose this as well.

**Review by apsdehal — COMMENTED** [review]
First set of comments. Will review rest of it some time later.

**Review by apsdehal — CHANGES_REQUESTED** [review]

File: mmf/models/mmf_transformer.py (hunk 1 above)
Comment (vedanuj) [reply mmf/models/mmf_transformer.py:28]: We only require the image encoder to be built here

File: mmf/models/mmf_transformer.py (hunk 2 above)
Comment (vedanuj) [reply mmf/models/mmf_transformer.py:31]: No, `Tensor` type works as it is.

File: mmf/models/mmf_transformer.py (hunk 3 above)
Comment (vedanuj) [reply mmf/models/mmf_transformer.py:46]: Yes, base transformer config. I would keep the name as is now.

File: mmf/models/mmf_transformer.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+    ):
+        super().__init__()
+        self.model_config = model_config
+        self.transformer_config = transformer_config
+        self._build_layers()
+        self._init_weights(transformer)
+
+    def _build_layers(self):
+        for modality in self.model_config.modalities:
+            if "text" in modality:
```
[the remark this answers is not in the document]
Comment (vedanuj) [reply mmf/models/mmf_transformer.py]: Modified in latest version

File: mmf/models/mmf_transformer.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+                    )
+                    .unsqueeze(0)
+                    .expand(input_ids[modality].size()[:2])
+                )
+            total_embedding += getattr(self, modality + "_pos_embedding")(
+                position_ids[modality]
+            )
+
+            if modality not in segment_ids:
+                segment_ids[modality] = torch.zeros(
```
[the remark this answers is not in the document]
Comment (vedanuj) [reply mmf/models/mmf_transformer.py]: Now we ask segment_id in the config, and not take any assumptions. If not defined in config, we just skip it.

File: mmf/models/mmf_transformer.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+                            self.transformer_config.hidden_size, eps=1e-12
+                        ),
+                    ),
+                )
+            # Set the position embeddings
+            setattr(
+                self,
+                modality + "_pos_embedding",
+                nn.Embedding(
+                    self.transformer_config.max_position_embeddings,
```
[the remark this answers is not in the document]
Comment (vedanuj) [reply mmf/models/mmf_transformer.py]: Added this to config

File: mmf/models/mmf_transformer.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+
+        # Token Type Embeddings
+        if hasattr(transformer.embeddings, "token_type_embeddings"):
+            token_vocab_size = self.transformer_config.type_vocab_size
+            self.token_type_embeddings.weight.data[:token_vocab_size].copy_(
+                transformer.embeddings.token_type_embeddings.weight
+            )
+            for idx in range(token_vocab_size, len(self.model_config.modalities)):
+                self.token_type_embeddings.weight.data[idx].copy_(
+                    transformer.embeddings.token_type_embeddings.weight.data.mean(dim=0)
```
[the remark this answers is not in the document]
Comment (vedanuj) [reply mmf/models/mmf_transformer.py:146]: Add normal random noise

File: mmf/configs/models/mmf_transformer/defaults.yaml (hunk 5 above)
Comment (vedanuj) [reply mmf/configs/models/mmf_transformer/defaults.yaml:1]: Added to base transformer class

**Review by apsdehal — APPROVED** [review]

**Pull request closed without merging**